File: AndroidApp/app/src/main/java/com/jamespfluger/alexadevicefinder/activities/LoginActivity.java
package com.jamespfluger.alexadevicefinder.activities;

import android.app.Activity;
import android.content.Intent;
import android.os.Bundle;
import android.view.View;
import android.widget.Button;
import android.widget.ProgressBar;
import android.widget.TextView;
import android.widget.Toast;

import com.amazon.identity.auth.device.AuthError;
import com.amazon.identity.auth.device.api.authorization.AuthCancellation;
import com.amazon.identity.auth.device.api.authorization.AuthorizationManager;
import com.amazon.identity.auth.device.api.authorization.AuthorizeListener;
import com.amazon.identity.auth.device.api.authorization.AuthorizeRequest;
import com.amazon.identity.auth.device.api.authorization.AuthorizeResult;
import com.amazon.identity.auth.device.api.authorization.ProfileScope;
import com.amazon.identity.auth.device.api.workflow.RequestContext;
import com.jamespfluger.alexadevicefinder.R;
import com.jamespfluger.alexadevicefinder.utilities.AmazonLoginHelper;

import java.util.Random;

public class LoginActivity extends Activity {
    private RequestContext requestContext;
    private Button loginButton;
    private TextView loginText;
    private ProgressBar loginProgressBar;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        // TODO: remove before release
        AmazonLoginHelper.signOut(getApplicationContext());
        requestContext = RequestContext.create(getApplicationContext());

        requestContext.registerListener(new AuthorizeListener() {
            @Override
            public void onSuccess(AuthorizeResult authorizeResult) {
                runOnUiThread(new Runnable() {
                    @Override
                    public void run() {
                        Toast.makeText(getApplicationContext(), "Successfully logged into Amazon.", Toast.LENGTH_SHORT).show();
                    }
                });
                switchToNameActivity();
            }

            @Override
            public void onError(AuthError authError) {
                runOnUiThread(new Runnable() {
                    @Override
                    public void run() {
                        Toast.makeText(getApplicationContext(), "Error logging in. Please try again.", Toast.LENGTH_SHORT).show();
                        setLoggingInState(false);
                    }
                });
            }

            @Override
            public void onCancel(AuthCancellation authCancellation) {
                runOnUiThread(new Runnable() {
                    @Override
                    public void run() {
                        Toast.makeText(getApplicationContext(), "Login cancelled.", Toast.LENGTH_SHORT).show();
                        setLoggingInState(false);
                    }
                });
            }
        });

        setContentView(R.layout.activity_login);
        initializeUI();
    }

    @Override
    protected void onResume() {
        super.onResume();
        requestContext.onResume();
    }

    private void initializeUI() {
        loginProgressBar = findViewById(R.id.loginProgressBar);
        loginButton = findViewById(R.id.loginButton);
        loginText = (TextView) findViewById(R.id.loginText);

        // TODO: implement way to pick between these
        Random r = new Random();
        if (r.nextBoolean()) {
            loginButton.setText("LOGIN WITH AMAZON");
            loginButton.setCompoundDrawablesRelativeWithIntrinsicBounds(R.drawable.amazon_logo, 0, 0, 0);
        } else {
            loginButton.setText("CONNECT WITH ALEXA");
            loginButton.setCompoundDrawablesRelativeWithIntrinsicBounds(R.drawable.amazon_alexa_logo, 0, 0, 0);
        }

        loginButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                AuthorizationManager.authorize(
                        new AuthorizeRequest.Builder(requestContext)
                                .addScopes(ProfileScope.userId())
                                .build()
                );
                setLoggingInState(true);
            }
        });
    }

    private void setLoggingInState(final boolean loggingIn) {
        if (loggingIn) {
            loginButton.setVisibility(Button.GONE);
            loginText.setVisibility(TextView.GONE);
            loginProgressBar.setVisibility(ProgressBar.VISIBLE);
        } else {
            loginButton.setVisibility(Button.VISIBLE);
            loginText.setVisibility(TextView.VISIBLE);
            loginProgressBar.setVisibility(ProgressBar.GONE);
        }
    }

    private void switchToNameActivity() {
        Intent otpIntent = new Intent(this, NameActivity.class);
        startActivity(otpIntent);
        finish();
    }
}

File: MobileApp/AlexaDeviceFinder/AlexaDeviceFinder.Android/Activities/LoginActivity.cs
using System;
using Android.App;
using Android.Content;
using Android.OS;
using Android.Views;
using Android.Widget;
using DeviceFinder.Droid.Utilities;

namespace DeviceFinder.Droid.Activities
{
    public class LoginActivity : Activity
    {
        //private RequestContext requestContext;
        private Button loginButton;
        private TextView loginText;
        private ProgressBar loginProgressBar;


        protected void onCreate(Bundle savedInstanceState)
        {
            base.OnCreate(savedInstanceState);
            // TODO: remove before release
            AmazonLoginHelper.SignOut(ApplicationContext);

            //requestContext = RequestContext.Create(ApplicationContext);
            //requestContext.RegisterListener(new AuthorizeListener());

            SetContentView(Resource.Layout.activity_login);
            initializeUI();
        }

        public void onSignOutSuccess()//AuthorizeResult authorizeResult)
        {
            Toast.MakeText(ApplicationContext, "Successfully logged into Amazon.", ToastLength.Short).Show();
            switchToActivity(typeof(NameActivity));
        }

        public void onSignOutError()//AuthError authError)
        {
            Toast.MakeText(ApplicationContext, "Error logging in. Please try again.", ToastLength.Short).Show();
            setLoggingInState(false);
        }

        public void onCancel()//AuthCancellation authCancellation)
        {
            Toast.MakeText(ApplicationContext, "Login cancelled.", ToastLength.Short).Show();
            setLoggingInState(false);
        }

        protected void onResume()
        {
            base.OnResume();
            //requestContext.OnResume();
        }

        private void initializeUI()
        {
            loginProgressBar = this.FindViewById(Resource.Id.loginProgressBar) as ProgressBar;
            loginButton = this.FindViewById(Resource.Id.loginButton) as Button;
            loginText = this.FindViewById(Resource.Id.loginText) as TextView;

            //
            loginButton.Text = ("LOGIN WITH AMAZON");
            loginButton.SetCompoundDrawablesRelativeWithIntrinsicBounds(Resource.Drawable.amazon_logo, 0, 0, 0);

            /* TODO: verify this is not needed
            loginButton.Text = ("CONNECT WITH ALEXA");
            loginButton.SetCompoundDrawablesRelativeWithIntrinsicBounds(R.Drawable.Amazon_alexa_logo, 0, 0, 0); */

            loginButton.Click += onLoginButtonClick; //.SetOnClickListener(new View.OnClickListener());
        }

        public void onLoginButtonClick(object sender, EventArgs args)
        {
            //AuthorizationManager.Authorize(
            //        new AuthorizeRequest.Builder(requestContext)
            //                .AddScopes(ProfileScope.UserId())
            //                .Build()
            //);
            setLoggingInState(true);
        }

        private void setLoggingInState(bool loggingIn)
        {
            if (loggingIn)
            {
                loginButton.Visibility = ViewStates.Gone;
                loginText.Visibility = ViewStates.Gone;
                loginProgressBar.Visibility = ViewStates.Visible;
            }
            else
            {
                loginButton.Visibility = ViewStates.Visible;
                loginText.Visibility = ViewStates.Visible;
                loginProgressBar.Visibility = ViewStates.Gone;
            }
        }

        private void switchToActivity(Type newActivity)
        {
            Intent newIntent = new Intent(this, newActivity);
            StartActivity(newIntent);
            Finish();
        }
    }
}